intercom/parse.go

```diff
@@ -0,0 +1,155 @@
+	if err != nil {
+		return nil, err
+	}
+
+	arr, err := jsonquery.New(node).Array(arrKey, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonquery.Convertor.ArrayToMap(arr)
+}
+
+func makeNextRecordsURL(reqLink *urlbuilder.URL) common.NextPageFunc {
+	return func(node *ajson.Node) (string, error) {
+		next, err := getNextPageStringURL(node)
+		if err == nil {
+			return next, nil
+		}
+
+		if !errors.Is(err, jsonquery.ErrNotString) {
+			// response from server doesn't meet any format that we expect
+			return "", err
+		}
+
+		// Probably, we are dealing with an object under `pages.next`
+		startingAfter, err := jsonquery.New(node, "pages", "next").Str("starting_after", true)
+		if err != nil {
+			return "", err
+		}
+
+		if startingAfter == nil {
+			// next page doesn't exist
+			return "", nil
+		}
+
+		reqLink.WithQueryParam("starting_after", *startingAfter)
+
+		return reqLink.String(), nil
+	}
+}
+
+// Some responses have full URL stored at `pages.next`.
+func getNextPageStringURL(node *ajson.Node) (string, error) {
+	nextPage, err := jsonquery.New(node, "pages").Str("next", true)
+	if err != nil {
+		return "", err
+	}
+
+	if nextPage == nil {
+		return "", nil
+	}
+
+	return *nextPage, nil
+}
+
+func getMarshaledData(records []map[string]interface{}, fields []string) ([]common.ReadResultRow, error) {
+	data := make([]common.ReadResultRow, len(records))
+
+	for i, record := range records {
+		data[i] = common.ReadResultRow{
+			Fields: common.ExtractLowercaseFieldsFromRaw(fields, record),
+			Raw:    record,
+		}
+	}
+
+	return data, nil
+}
+
+// The key that stores array in response payload will be dynamically figured out.
+// Ex: {"data": []} vs {"teams":[]} vs {"segments":[]}.
+func extractListFieldName(node *ajson.Node) (string, error) {
+	// default field at which list is stored
+	defaultFieldName := "data"
+
+	fieldName, err := jsonquery.New(node).Str("type", true)
+	if err != nil {
+		return "", err
+	}
+
+	if fieldName == nil {
+		// this object has no `type` field to infer where the array is situated
+		// it is unexpected to encounter it
+		return defaultFieldName, nil
+	}
+
+	name := *fieldName
+	// by applying plural form to the object name we will the name of field containing array
+	// Ex with `list` suffix:
+	// 		activity_log.list => activity_logs
+	// 		admin.list => admins
+	// 		conversation.list => conversations
+	// 		segment.list => segments
+	// 		team.list => teams
+	// Exceptions:
+	//		event.summary => events
+
+	parts := strings.Split(name, ".")
+	if len(parts) == 2 { // nolint:gomnd
+		// custom name is used when it has 2 parts
+		return applyPluralForm(parts[0]), nil
+	}
+
+	// usually when we have a pure `list` type it means array is stored at `data` field
+	return defaultFieldName, nil
+}
+
+func applyPluralForm(word string) string {
+	// The requirement for Intercom's object pluralisation is low and simple
+	// There is no need to consider all English orthography cases
+	return word + "s"
```

Comment:
Adding single `s` is significant for all existing endpoints under Intercom. Maybe later we would need to handle `es`, or words ending with `y`, some sort of library would help.

Reply:
If it's easier you could use something like this: https://github.com/gertd/go-pluralize

Generally speaking, dependencies on 3rd party libraries are ok, but the license has to be closed-source friendly (MIT, BSD, Apache, etc).

Reply:
Cool, I added this library. https://github.com/amp-labs/connectors/pull/346/commits/62e51e2860e7e16ac5079bbd38fe15764b212f5b
Tests pass.